File: ios/pqp/Sources/Chat/DayLabels.swift
import Foundation

/// The day rows of a transcript, kept the same as the web's `formatDayLabel`
/// in `client/src/lib/utils.ts`.
///
/// The rule there: "Today", "Yesterday", the weekday for anything inside the
/// last seven days, and a short date beyond that with the year only when it is
/// not this year. Pure, and reading the calendar and the clock from the caller,
/// so the one place it can be wrong (a message sent at 23:58 read at 00:02) is
/// pinned by a test rather than by whoever is awake at midnight.
enum DayLabels {
    /// Whether two instants fall on the same calendar day for this reader.
    static func isSameDay(_ a: Date, _ b: Date, calendar: Calendar = .current) -> Bool {
        calendar.isDate(a, inSameDayAs: b)
    }

    static func label(
        for date: Date,
        now: Date = Date(),
        calendar: Calendar = .current,
        locale: Locale = .current
    ) -> String {
        let start = calendar.startOfDay(for: date)
        let today = calendar.startOfDay(for: now)
        let daysAgo = calendar.dateComponents([.day], from: start, to: today).day ?? 0

        if daysAgo == 0 { return String(localized: "Today") }
        if daysAgo == 1 { return String(localized: "Yesterday") }

        // `< 7`, not `2...6`: the web spells out the weekday for anything less
        // than a week old, which includes a clock slightly ahead of the server's.
        // Fields, not a preset: `.abbreviated` would carry the year every time,
        // and the year is the one part that is conditional.
        var style = Date.FormatStyle(date: .omitted, time: .omitted, locale: locale, calendar: calendar)
            .month(.abbreviated)
            .day()
        if daysAgo < 7 { style = style.weekday(.wide) }
        let sameYear = calendar.component(.year, from: date) == calendar.component(.year, from: now)
        if !sameYear { style = style.year() }
        let text = date.formatted(style)
        // Portuguese weekday and month names are lowercase in CLDR. A row that
        // stands on its own reads better with a capital, the way the web's
        // rendering of the same string does.
        return text.prefix(1).uppercased(with: locale) + text.dropFirst()
    }
}

File: android/app/src/main/kotlin/gg/pqp/app/ui/chat/DayLabels.kt
package gg.pqp.app.ui.chat

import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import java.util.Locale

/**
 * What the separator between two calendar days says.
 *
 * `Today` and `Yesterday` are words, so they come back as cases for the screen
 * to look up in resources; everything older is a date the platform can format
 * for the reader's locale, so it comes back already spelled out.
 */
sealed interface DayLabel {
    data object Today : DayLabel
    data object Yesterday : DayLabel
    data class Dated(val text: String) : DayLabel
}

/**
 * The day rows of a transcript, kept the same as the web's `formatDayLabel`
 * in `client/src/lib/utils.ts`.
 *
 * The rule there: "Today", "Yesterday", the weekday for anything inside the
 * last seven days, and a short date beyond that with the year only when it is
 * not this year. Pure, and reading the zone and the clock from the caller, so
 * the one place it can be wrong (a message sent at 23:58 read at 00:02) is
 * pinned by a test rather than by whoever is awake at midnight.
 */
object DayLabels {

    /** The calendar day a server timestamp lands on for this reader. */
    fun localDate(iso: String, zone: ZoneId): LocalDate? =
        runCatching { Instant.parse(iso).atZone(zone).toLocalDate() }.getOrNull()

    /**
     * Whether two timestamps fall on the same local day. False on a timestamp
     * that will not parse, which draws one separator too many rather than
     * silently gluing two days together.
     */
    fun isSameDay(a: String, b: String, zone: ZoneId = ZoneId.systemDefault()): Boolean {
        val first = localDate(a, zone) ?: return false
        val second = localDate(b, zone) ?: return false
        return first == second
    }

    fun labelFor(
        iso: String,
        today: LocalDate = LocalDate.now(),
        zone: ZoneId = ZoneId.systemDefault(),
        locale: Locale = Locale.getDefault(),
    ): DayLabel? {
        val date = localDate(iso, zone) ?: return null
        val daysAgo = ChronoUnit.DAYS.between(date, today)
        return when (daysAgo) {
            0L -> DayLabel.Today
            1L -> DayLabel.Yesterday
            else -> DayLabel.Dated(
                format(
                    date = date,
                    // `< 7`, not `in 2..6`: the web spells out the weekday for
                    // anything less than a week old, which includes a clock
                    // that is slightly ahead of the server's.
                    withWeekday = daysAgo < 7,
                    withYear = date.year != today.year,
                    locale = locale,
                ),
            )
        }
    }

    /**
     * `toLocaleDateString` with `{ weekday, month: "short", day }` is what the
     * web renders, and there is no skeleton API in `java.time`, so the two
     * shapes the app ships are spelled out per language. Portuguese puts "de"
     * between the parts; everything else reads month first, the way the
     * English CLDR data does.
     */
    private fun format(date: LocalDate, withWeekday: Boolean, withYear: Boolean, locale: Locale): String {
        val portuguese = locale.language.equals("pt", ignoreCase = true)
        val pattern = buildString {
            if (withWeekday) append("EEEE, ")
            append(if (portuguese) "d 'de' MMM" else "MMM d")
            if (withYear) append(if (portuguese) " 'de' yyyy" else ", yyyy")
        }
        val text = DateTimeFormatter.ofPattern(pattern, locale).format(date)
        // Portuguese weekday and month names are lowercase in CLDR. A row that
        // stands on its own reads better with a capital, the way the web's
        // rendering of the same string does.
        return text.replaceFirstChar { if (it.isLowerCase()) it.titlecase(locale) else it.toString() }
    }
}
